extensions/ql-vscode/src/common/selection-commands.ts

```diff
@@ -0,0 +1,51 @@
+import { showAndLogErrorMessage } from "../helpers";
+import {
+  ExplorerSelectionCommandFunction,
+  TreeViewContextMultiSelectionCommandFunction,
+  TreeViewContextSingleSelectionCommandFunction,
+} from "./commands";
+
+// A hack to match types that are not an array, which is useful to help avoid
+// misusing createSingleSelectionCommand, e.g. where T accidentally gets instantiated
+// as DatabaseItem[] instead of DatabaseItem.
+type NotArray = object & { length?: never };
+
+// A way to get the type system to help assert that one type is a supertype of another.
+type CreateSupertypeOf<Super, Sub extends Super> = Sub;
+
+// This asserts that SelectionCommand is assignable to all of the different types of
+// SelectionCommand defined in commands.ts. The intention is the output from the helpers
+// in this file can be used with any of the select command types and cna handle any of
+// the inputs.
+type SelectionCommand<T extends NotArray> = CreateSupertypeOf<
+  TreeViewContextMultiSelectionCommandFunction<T> &
+    TreeViewContextSingleSelectionCommandFunction<T> &
+    ExplorerSelectionCommandFunction<T>,
+  (singleItem: T, multiSelect?: T[] | undefined) => Promise<void>
+>;
+
+export function createSingleSelectionCommand<T extends NotArray>(
+  f: (argument: T) => Promise<void>,
+  itemName: string,
+): SelectionCommand<T> {
+  return async (singleItem, multiSelect) => {
+    if (multiSelect === undefined || multiSelect.length === 1) {
+      return f(singleItem);
+    } else {
+      void showAndLogErrorMessage(`Please select a single ${itemName}.`);
+      return;
+    }
+  };
+}
+
+export function createMultiSelectionCommand<T extends NotArray>(
+  f: (argument: T[]) => Promise<void>,
+): SelectionCommand<T> {
+  return async (singleItem, multiSelect) => {
+    if (multiSelect !== undefined && multiSelect.length > 0) {
+      return f(multiSelect);
+    } else {
+      return f([singleItem]);
```

Comment:
As per the types and my experimentation, `singleItem` cannot be undefined. If you can find a way to trigger one of our commands and get `singleItem` to be undefined I'd be very interested to hear. Otherwise I'd prefer to keep our types and code simple.

Reply:
We seem to be checking whether a databaseItem is `undefined` in three places?

https://github.com/github/vscode-codeql/pull/2287/commits/40f309224457cc886a4e04e69e80cf25c1ef2e23#diff-b11b6f7e1d76e818f721c94ae740f046663c80d519c2f515f2f78344f90e6c78R518

https://github.com/github/vscode-codeql/pull/2287/commits/40f309224457cc886a4e04e69e80cf25c1ef2e23#diff-b11b6f7e1d76e818f721c94ae740f046663c80d519c2f515f2f78344f90e6c78R540

https://github.com/github/vscode-codeql/pull/2287/commits/40f309224457cc886a4e04e69e80cf25c1ef2e23#diff-b11b6f7e1d76e818f721c94ae740f046663c80d519c2f515f2f78344f90e6c78R564

Reply:
Thanks for pointing those out. The commands around upgrading databases are a little complicated because we kinda have two notions of being "selected". You can click to highlight entries in the list (i.e. the vscode tree view), and this state gets passed through to the `codeQLDatabases.upgradeDatabase` command if you select to upgrade the databases from the context menu. You can also click "make selected" on a database and this state is managed internally, and this is what is used for the `codeQL.upgradeCurrentDatabase` or `codeQL.runQuery` (a.k.a. "Run Query on Selected Database") commands.

Looking at each bit of code:
- https://github.com/github/vscode-codeql/commit/40f309224457cc886a4e04e69e80cf25c1ef2e23#diff-b11b6f7e1d76e818f721c94ae740f046663c80d519c2f515f2f78344f90e6c78R518
  - As mentioned above, the `codeQL.upgradeCurrentDatabase` command isn't really a "selection command" like the ones we're looking at in this PR. It's operating on the currently selected local database, but it's not a vscode tree view selection like the other cases in this PR. It's a selection that we manage ourselves, and thus you can trigger this command at any time to operate on the "current" database, and there may be a database selected or not.
- https://github.com/github/vscode-codeql/commit/40f309224457cc886a4e04e69e80cf25c1ef2e23#diff-b11b6f7e1d76e818f721c94ae740f046663c80d519c2f515f2f78344f90e6c78R540
  - This commit is from quite early in the PR and is doing some refactoring before we get into the rest of the PR. The code being linked to is already changed on the latest commit.
- https://github.com/github/vscode-codeql/commit/40f309224457cc886a4e04e69e80cf25c1ef2e23#diff-b11b6f7e1d76e818f721c94ae740f046663c80d519c2f515f2f78344f90e6c78R564
  - Thanks. This check is unnecessary. It's an oversight from me just wrapping the existing function implementation with `await Promise.all`. I've pushed a commit to remove this check.